lib/node_modules/@stdlib/repl/lib/eager_evaluator.js

```diff
@@ -0,0 +1,330 @@
+var inspect = require( 'util' ).inspect;
+var logger = require( 'debug' );
+var parse = require( 'acorn' ).parse;
+var replace = require( '@stdlib/string/replace' );
+var setNonEnumerableReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
+var deepCopy = require( '@stdlib/utils/copy' );
+var max = require( '@stdlib/math/base/special/max' );
+var processCommand = require( './process_command.js' );
+var compileCommand = require( './compile_command.js' );
+var ANSI_COLORS = require( './ansi_colors.js' );
+
+
+// VARIABLES //
+
+var debug = logger( 'repl:eager-evaluator' );
+var AOPTS = {
+	'ecmaVersion': 'latest'
+};
+var ROPTS = {
+	'timeout': 100, // (in milliseconds) shouldn't be much noticeable when naturally typing
+	'displayErrors': false,
+	'breakOnSigint': true // Node.js >=6.3.0
+};
+var tempDB = {
+	'base_sin': {
+		'isPure': true
+	}
+};
+var ANSI_GRAY = ANSI_COLORS[ 'brightBlack' ];
+var ANSI_RESET = ANSI_COLORS[ 'reset' ];
+
+
+// MAIN //
+
+/**
+* Constructor for creating an eager evaluator.
+*
+* @private
+* @param {REPL} repl - repl instance
+* @param {Object} rli - readline instance
+* @param {Boolean} enabled - boolean indicating whether the eager evaluator should be initially enabled
+* @returns {EagerEvaluator} eager evaluator instance
+*/
+function EagerEvaluator( repl, rli, enabled ) {
+	if ( !(this instanceof EagerEvaluator) ) {
+		return new EagerEvaluator( repl, rli, enabled );
+	}
+	debug( 'Creating a new eager evaluator...' );
+
+	// Cache a reference to the provided REPL instance:
+	this._repl = repl;
+
+	// Cache a reference to the readline interface:
+	this._rli = rli;
+
+	// Cache a reference to the command array:
+	this._cmd = repl._cmd;
+
+	// Initialize a flag indicating whether the eager evaluator is enabled:
+	this._enabled = enabled;
+
+	// Initialize a flag indicating whether we are currently previewing eagerly-evaluated output:
+	this._isPreviewing = false;
+
+	return this;
+}
+
+/**
+* Checks if the code is side-effect-free.
+*
+* @private
+* @name _isSideEffectFree
+* @memberof EagerEvaluator.prototype
+* @type {Function}
+* @param {string} code - input code
+* @returns {boolean} - boolean indicating whether the code is side-effect-free
+*
+*/
+setNonEnumerableReadOnly( EagerEvaluator.prototype, '_isSideEffectFree', function isSideEffectFree( code ) {
+	var ast;
+	var i;
+
+	try {
+		ast = parse( code, AOPTS );
+	}
+	catch ( err ) {
+		debug( 'Encountered an error when generating AST: %s', err.message );
+		return false;
+	}
+	// Iterate from each node in the body:
+	for ( i = 0; i < ast.body.length; i++ ) {
+		if ( !traverse( ast.body[ i ] ) ) {
+			return false;
+		}
+	}
+	return true;
+
+	/**
+	* Recursively traverses the node to determine whether the node is side-effect-free or not.
+	*
+	* @private
+	* @param {Object} node - ast node
+	* @returns {Boolean} - boolean indicating whether the node is side-effect-free
+	*/
+	function traverse( node ) {
```

Comment:
Can this function be moved to the parent scope? I don't believe that the function's implementation depends on anything from the `isSideEffectFree` context.